src/System.Net.Security/src/System/Net/_SecureChannel.cs

```diff
@@ -0,0 +1,1487 @@
+            securityBuffer[1] = new SecurityBuffer(e_writeBuffer, _headerSize, size, SecurityBufferType.Data);
+            securityBuffer[2] = new SecurityBuffer(e_writeBuffer, _headerSize + size, _trailerSize, SecurityBufferType.Trailer);
+            securityBuffer[3] = new SecurityBuffer(null, SecurityBufferType.Empty);
+
+            int errorCode = SSPIWrapper.EncryptMessage(GlobalSSPI.SSPISecureChannel, _securityContext, securityBuffer, 0);
+
+            if (errorCode != 0)
+            {
+                GlobalLog.Leave("SecureChannel#" + Logging.HashString(this) + "::Encrypt ERROR", errorCode.ToString("x"));
+                return (Interop.SecurityStatus)errorCode;
+            }
+            else
+            {
+                output = e_writeBuffer;
+
+                // The full buffer may not be used.
+                resultSize = securityBuffer[0].size + securityBuffer[1].size + securityBuffer[2].size;
+                GlobalLog.Leave("SecureChannel#" + Logging.HashString(this) + "::Encrypt OK", "data size:" + resultSize.ToString());
+                return Interop.SecurityStatus.OK;
+            }
+        }
+
+        internal Interop.SecurityStatus Decrypt(byte[] payload, ref int offset, ref int count)
+        {
+            GlobalLog.Print("SecureChannel#" + Logging.HashString(this) + "::Decrypt() - offset: " + offset.ToString() + " size: " + count.ToString() + " buffersize: " + payload.Length.ToString());
+
+            if (offset < 0 || offset > (payload == null ? 0 : payload.Length))
+            {
+                GlobalLog.Assert(false, "SecureChannel#" + Logging.HashString(this) + "::Encrypt", "Argument 'offset' out of range.");
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > (payload == null ? 0 : payload.Length - offset))
+            {
+                GlobalLog.Assert(false, "SecureChannel#" + Logging.HashString(this) + "::Encrypt", "Argument 'count' out of range.");
+                throw new ArgumentOutOfRangeException("count");
+            }
+
+            // Decryption using SCHANNEL requires four buffers.
+            SecurityBuffer[] decspc = new SecurityBuffer[4];
+            decspc[0] = new SecurityBuffer(payload, offset, count, SecurityBufferType.Data);
+            decspc[1] = new SecurityBuffer(null, SecurityBufferType.Empty);
+            decspc[2] = new SecurityBuffer(null, SecurityBufferType.Empty);
+            decspc[3] = new SecurityBuffer(null, SecurityBufferType.Empty);
+
+            Interop.SecurityStatus errorCode = (Interop.SecurityStatus)SSPIWrapper.DecryptMessage(GlobalSSPI.SSPISecureChannel, _securityContext, decspc, 0);
+
+            count = 0;
+            for (int i = 0; i < decspc.Length; i++)
+            {
+                // Successfully decoded data and placed it at the following position in the buffer,
+                if ((errorCode == Interop.SecurityStatus.OK && decspc[i].type == SecurityBufferType.Data)
```

Comment:
Personally, I'd try to reduce the line length and complexity here by doing this like
```C#
// Useful comment here
SecurityBufferType desiredType =
    (errorCode == Interop.SecurityStatus.OK) ? SecurityBufferType.Data : SecurityBufferType.Extra;

count = 0;

for (int i = 0; i < decspc.Length; i++)
{
    if (decspc[i].type == desiredType)
    {
        // body
    }
}
```